tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp

```diff
@@ -0,0 +1,153 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace TestHelpers::TensorExpressions {
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Test that evaluating a right hand side tensor expression containing a
+/// single rank 2 tensor correctly assigns the data to the evaluated left hand
+/// side tensor
+///
+/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
+/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
+/// denote the ordering of the generic indices of the RHS tensor expression. In
+/// the RHS tensor expression, it means `TensorIndexA` is the first index used
+/// and `TensorIndexB` is the second index used.
+///
+/// If we consider the RHS tensor's generic indices to be (a, b), then this test
+/// checks that the data in the evaluated LHS tensor is correct according to the
+/// index orders of the LHS and RHS. The two possible cases that are checked are
+/// when the LHS tensor is evaluated with index order (a, b) and when it is
+/// evaluated with the index order (b, a).
+///
+/// \tparam DataType the type of data being stored in the Tensors
+/// \tparam RhsSymmetry the ::Symmetry of the RHS Tensor
+/// \tparam RhsTensorIndexTypeList the RHS Tensor's typelist of
+/// \ref SpacetimeIndex "TensorIndexType"s
+/// \param tensorindex_a the first TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_a`
+/// \param tensorindex_b the second TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_B`
+template <typename DataType, typename RhsSymmetry,
+          typename RhsTensorIndexTypeList, typename TensorIndexA,
+          typename TensorIndexB>
+void test_evaluate_rank_2_core(const TensorIndexA& tensorindex_a,
+                               const TensorIndexB& tensorindex_b) noexcept {
+  Tensor<DataType, RhsSymmetry, RhsTensorIndexTypeList> R_ab(5_st);
+  std::iota(R_ab.begin(), R_ab.end(), 0.0);
+
+  const size_t dim_a = tmpl::at_c<RhsTensorIndexTypeList, 0>::dim;
+  const size_t dim_b = tmpl::at_c<RhsTensorIndexTypeList, 1>::dim;
+
+  // L_{ab} = R_{ab}
+  const auto L_ab = ::TensorExpressions::evaluate<TensorIndexA, TensorIndexB>(
```

Comment:
I think this is a great suggestion. Last night I was working on the contraction code and I ran into a bug like the one you're mentioning, and this would've been caught earlier by my contraction tests if I hadn't been using `auto`, so today I was going to add asserts to check the indices on the LHS - but I like this much better. Thanks!

Reply:
Something I didn't notice at first though is that if I'm to do this for all of the cases in the tests, then instead of
```
const Tensor<DataType, RhsSymmetry, tmpl::list<TensorIndexA, TensorIndexB>> LHS = ...
```
it will have to be
```
const Tensor<DataType, <the LHS reordered symmetry>, <the LHS reordered tensor index type list>> LHS = ...
```
so that the LHS `Tensor` has its symmetry and tensor index type list reordered the same way as the `TensorIndex`s are being reordered (essentially, what the struct in `Evaluate.hpp` currently named `LhsTensor` does)

Reply:
Oh good! It might be helpful to add a short comment (just once per file the first time you replace `auto` with the expected type) saying something like `// Use explicit type (vs auto) so the compiler checks the return type of the TensorExpression` ? That might help avoid someone thinking they can simplify the code in the future by putting `auto` back in 😛

Reply:
Done! The code got decently more verbose for the rank 3 and rank 4 `impl` (previously, `core`) functions, so let me know if you would prefer me to handle this in a different way than I did in 317e87d

Reply:
I agree this did get a bit verbose, but it does the checks we wanted. I'm happy with it!

I also can't think of an alternative way of doing these checks that isn't equally verbose, because regardless of how you structure things, you're going to need to write out the expected symmetry and index list – and that's the most cumbersome part.